cocotb/cocotb.bzl

```diff
@@ -0,0 +1,267 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Rules for running tests using Cocotb framework"""
+
+load("//verilog:providers.bzl", "VerilogInfo")
+load("@rules_python//python:defs.bzl", "PyInfo")
+
+def _list_to_argstring(data, argname, attr = None, operation = None):
+    result = " --{}".format(argname) if data else ""
+    for value in data:
+        elem = value if attr == None else getattr(value, attr)
+        elem = elem if operation == None else operation(elem)
+        result += " {}".format(elem)
+    return result
+
+def _dict_to_argstring(data, argname):
+    result = " --{}".format(argname) if data else ""
+    for key, value in data.items():
+        result += " {}={}".format(key, value)
+    return result
+
+def _files_to_argstring(data, argname):
+    return _list_to_argstring(data, argname, "path")
+
+def _pymodules_to_argstring(data, argname):
+    remove_py = lambda s: s.removesuffix(".py")
+    return _list_to_argstring(data, argname, "basename", remove_py)
+
+def _remove_duplicates_from_list(data):
+    result = []
+    for e in data:
+        if e not in result:
+            result.append(e)
+    return result
+
+def _collect_verilog_files(ctx):
+    # from VerilogInfo
+    transitive_srcs_list = [
+        dep
+        for dep in ctx.attr.verilog_sources
+        if VerilogInfo in dep
+    ]
+    transitive_srcs_depset = depset(
+        [],
+        transitive = [dep[VerilogInfo].dag for dep in transitive_srcs_list],
+    )
+    verilog_srcs = [
+        verilog_info_struct.srcs
+        for verilog_info_struct in transitive_srcs_depset.to_list()
+    ]
+    verilog_files = [src for sub_tuple in verilog_srcs for src in sub_tuple]
+
+    # from normal files
+    verilog_files += ctx.files.verilog_sources
```

Comment:
This can be slow with a large number of files consider using a depset.

https://bazel.build/rules/lib/builtins/depset

Reply:
Let me know if the current changes are what you had in mind

Reply:
LGTM.